pkg/agent/tcpconn/dropper.go

```diff
@@ -0,0 +1,48 @@
+package tcpconn
+
+import (
+	"fmt"
+	"hash/crc32"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+// Dropper is an interface implemented by objects that decide whether a packet should be dropped (rejected).
+type Dropper interface {
+	Drop(packetBytes []byte) bool
+}
+
+// TCPConnectionDropper is a dropper that drops a defined percentage of TCP the connections it sees.
+type TCPConnectionDropper struct {
+	DropRate float64
+	// TODO: Add DropPeriod.
+}
+
+// Drop decides whether a packet should be dropped by taking the modulus of hash of the connection it belongs to and
+// comparing it to a threshold derived from DropRate.
+func (tcd TCPConnectionDropper) Drop(packetBytes []byte) bool {
+	packet := gopacket.NewPacket(packetBytes, layers.LayerTypeIPv4, gopacket.Default)
+
+	ipLayer := packet.Layer(layers.LayerTypeIPv4)
+	if ipLayer == nil {
+		return false
+	}
+	ip, _ := ipLayer.(*layers.IPv4)
+
+	tcpLayer := packet.Layer(layers.LayerTypeTCP)
+	if tcpLayer == nil {
+		return false
+	}
+	tcp, _ := tcpLayer.(*layers.TCP)
+
+	// fourTuple uniquely identifies this connection by its 4-tuple: source address and port, and destination address
+	// and port.
+	fourTuple := fmt.Sprintf("%v:%d:%v:%d", ip.SrcIP, tcp.SrcPort, ip.DstIP, tcp.DstPort)
```

Comment:
Maybe we could be more efficient here and just copy the files into a `[]byte` using low-level manipulations.

Reply:
Good point, with this being in the hot path it is worth trying, specially as we know exactly the sizes of those fields so we can operate entirely on the stack.

It will make the code more verbose though, but seems worth trying.

Reply:
> It will make the code more verbose though, but seems worth trying

I think we could consider this in the context of the other optimization I mentioned in my general comment to the PR.

Reply:
I ran some benchmarks using a binary implementation as follows:
```diff
diff --git a/pkg/agent/tcpconn/dropper.go b/pkg/agent/tcpconn/dropper.go
index a3e5bfa..1c3cf69 100644
--- a/pkg/agent/tcpconn/dropper.go
+++ b/pkg/agent/tcpconn/dropper.go
@@ -1,7 +1,7 @@
 package tcpconn
 
 import (
-	"fmt"
+	"encoding/binary"
 	"hash/crc32"
 
 	"github.com/google/gopacket"
@@ -35,13 +35,30 @@ func (tcd TCPConnectionDropper) Drop(packetBytes []byte) bool {
 	}
 	tcp, _ := tcpLayer.(*layers.TCP)
 
-	// fourTuple uniquely identifies this connection by its 4-tuple: source address and port, and destination address
-	// and port.
-	fourTuple := fmt.Sprintf("%v:%d:%v:%d", ip.SrcIP, tcp.SrcPort, ip.DstIP, tcp.DstPort)
+	ftBuf := make([]byte, 36)
+	fourTuple(ftBuf, ip, tcp)
 
 	hash := crc32.NewIEEE()
-	_, _ = hash.Write([]byte(fourTuple))
+	_, _ = hash.Write(ftBuf)
 	checksum := hash.Sum32()
 
 	return (checksum % 100) < uint32(100*tcd.DropRate)
 }
+
+// 4 tuple writes the 4-tuple of a given packet in to dst, given its ip and tcp layers.
+// dst must be at least 36 bytes long.
+func fourTuple(dst []byte, ip *layers.IPv4, tcp *layers.TCP) {
+	offset := 0
+
+	// Go's ip.IP representation can always be 16 bytes, even for v4 addresses.
+	copy(dst[offset:], ip.SrcIP)
+	offset += 16
+
+	copy(dst[offset:], ip.DstIP)
+	offset += 16
+
+	binary.LittleEndian.PutUint16(dst[offset:], uint16(tcp.SrcPort))
+	offset += 2
+
+	binary.LittleEndian.PutUint16(dst[offset:], uint16(tcp.DstPort))
+}
```

The results did not seem significant to me. I would guess that the biggest performance hit is in the context switch and copy operation from kernel to userspace.


With printf:
```
goos: linux
goarch: amd64
pkg: github.com/grafana/xk6-disruptor/pkg/agent/tcpconn
cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	5603779208 ns/op	      3390 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4944936075 ns/op	      5082 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4759099532 ns/op	      5082 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4766988543 ns/op	      3397 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4612135500 ns/op	      3386 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4725495315 ns/op	      5091 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4749048971 ns/op	      3391 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4719424771 ns/op	      5068 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4548333090 ns/op	      3385 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4860284601 ns/op	      5090 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11216304335 ns/op	       159.4 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10939411358 ns/op	       154.6 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	12255563572 ns/op	       132.5 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10730306312 ns/op	       162.0 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10933200240 ns/op	       159.4 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10622811473 ns/op	       162.0 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10899302583 ns/op	       154.6 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	12439242417 ns/op	       132.5 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11246890227 ns/op	       152.3 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11371617435 ns/op	       152.3 MiB/s
PASS
ok  	github.com/grafana/xk6-disruptor/pkg/agent/tcpconn	174.995s
```

With binary:
```
goos: linux
goarch: amd64
pkg: github.com/grafana/xk6-disruptor/pkg/agent/tcpconn
cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	5133993561 ns/op	      5091 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4450578743 ns/op	      5049 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4571364588 ns/op	      3397 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4530964040 ns/op	      5084 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4826679764 ns/op	      5072 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4610344644 ns/op	      5071 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4663539509 ns/op	      5092 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4648757006 ns/op	      3395 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4772370737 ns/op	      3396 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4614075378 ns/op	      3393 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11090325193 ns/op	       152.3 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11146554238 ns/op	       150.1 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10071293977 ns/op	       175.9 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10559728677 ns/op	       167.3 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11277076783 ns/op	       147.9 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11470568318 ns/op	       145.6 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11209702655 ns/op	       152.3 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10535403751 ns/op	       170.1 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11258844511 ns/op	       147.9 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10295333791 ns/op	       172.9 MiB/s
PASS
ok  	github.com/grafana/xk6-disruptor/pkg/agent/tcpconn	171.685s
```

Reply:
I agree that the main performance penalty is from processing of packages in user space. 
I didn't expect any significant change only by doing this modification, hence my previous comment:
> I think we could consider this in the context of the other optimization I mentioned in my general comment to the PR.

Reply:
For completeness, I ran this benchmark with the most naive implementation possible, just using the TCP src port for hashing:

```go
// Drop decides whether a packet should be dropped by taking the modulus of hash of the connection it belongs to and
// comparing it to a threshold derived from DropRate.
func (tcd TCPConnectionDropper) Drop(packetBytes []byte) bool {
	packet := gopacket.NewPacket(packetBytes, layers.LayerTypeIPv4, gopacket.Default)

	tcpLayer := packet.Layer(layers.LayerTypeTCP)
	if tcpLayer == nil {
		return false
	}
	tcp, _ := tcpLayer.(*layers.TCP)

	checksum := uint16(tcp.SrcPort)

	return (checksum % 100) < uint16(100*tcd.DropRate)
}
```


Results seem very similar:
```
goos: linux
goarch: amd64
pkg: github.com/grafana/xk6-disruptor/pkg/agent/tcpconn
cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	5249395289 ns/op	      5054 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4540443247 ns/op	      5086 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4508760796 ns/op	      5079 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4616980534 ns/op	      3388 MiB/s
Benchmark_DisruptorThroughput/disruption=agent-disabled-16         	       1	4609052036 ns/op	      3388 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11066621773 ns/op	       152.3 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10831065757 ns/op	       159.5 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11192662357 ns/op	       150.1 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	11270983583 ns/op	       157.0 MiB/s
Benchmark_DisruptorThroughput/disruption=accept-all-16             	       1	10230564251 ns/op	       173.0 MiB/s
PASS
ok  	github.com/grafana/xk6-disruptor/pkg/agent/tcpconn	86.457s
```

I'll continue trying your other suggestions to see where they leave us.

Reply:
I've left the native 4tuple calculation on a branch here for future reference: https://github.com/grafana/xk6-disruptor/commit/515bdd9bbcbfa4bca28da9442db05020f0e6585f

For now I'd suggest we leave performance optimization to a follow-up PR.